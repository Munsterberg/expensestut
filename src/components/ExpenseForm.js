import React from 'react';
import moment from 'moment';
import { func, object } from 'prop-types';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }

  static propTypes = {
    onSubmit: func,
    expense: object
  };

  _onInputChange = (e) => {
    const element = e.target;
    const stateObj = {};

    stateObj[element.name] = element.value;
    this.setState(() => (stateObj));
  }

  _onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }

  _onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  }

  _onCalendarFocus = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  }

  _onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please submit both description and an amount!' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  }

  render() {
    const { error, description, note, amount, createdAt, calendarFocused } = this.state;
    return (
      <div>
        {
          error && <p>{ error }</p>
        }
        <form onSubmit={this._onSubmit}>
          <input
            type="text"
            placeholder="Description"
            name="description"
            value={description}
            onChange={this._onInputChange}
            autoFocus
          />
          <input
            type="text"
            placeholder="Amount"
            value={amount}
            onChange={this._onAmountChange}
          />
          <textarea
            placeholder="Add a note...(optional)"
            name="note"
            value={note}
            onChange={this._onInputChange}
          >
          </textarea>
          <SingleDatePicker
            date={createdAt}
            onDateChange={this._onDateChange}
            focused={calendarFocused}
            onFocusChange={this._onCalendarFocus}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
