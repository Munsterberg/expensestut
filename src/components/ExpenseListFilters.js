import React from 'react';
import { connect } from 'react-redux';
import { object, func } from 'prop-types';
import { DateRangePicker } from 'react-dates';

import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };

  static propTypes = {
    filters: object,
    setTextFilter: func,
    setStartDate: func,
    setEndDate: func,
    sortByDate: func,
    sortByAmount: func
  };

  _onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }

  _onFocusChange = (focusedInput) => {
    this.setState(() => ({ calendarFocused: focusedInput }));
  }

  _onFilterChange = (e) => {
    this.props.setTextFilter(e.target.value);
  }

  _onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  }

  render() {
    return (
      <div>
        <h4>Filter Expenses</h4>
        <input type="text" value={this.props.filters.text} onChange={this._onFilterChange} />
        <select
          value={this.props.filters.sortBy}
          onChange={this._onSortChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this._onDatesChange}
          focusedInput={this.state.calendarFocused}
          numberOfMonths={1}
          onFocusChange={this._onFocusChange}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
