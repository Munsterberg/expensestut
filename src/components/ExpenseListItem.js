import React from 'react';
import { string, number, func } from 'prop-types';
import { Link } from 'react-router-dom';

ExpenseListItem.propTypes = {
  description: string,
  amount: number,
  createdAt: number,
  id: string,
  dispatch: func
};

function ExpenseListItem({ id, description, amount, createdAt }) {
  return (
    <div>
      <Link to={`/edit/${id}`}><h3>{ description }</h3></Link>
      <p>{amount} - {createdAt}</p>
    </div>
  );
}

export default ExpenseListItem;
