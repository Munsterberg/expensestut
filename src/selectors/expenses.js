const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {
    const startDateMatch = typeof startDate !== 'number' || expenses.createdAt >= startDate;
    const endDateMatch = typeof endDate !== 'number' || expenses.createdAt <= endDate;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
  });
};

export default getVisibleExpenses;
