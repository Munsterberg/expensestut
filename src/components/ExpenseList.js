import React from 'react';
import { connect } from 'react-redux';
import { array } from 'prop-types';

import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../selectors/expenses';

const requiredProps = {
  expenses: array
};

export const ExpenseList = (props) => {
  return (
    <div>
      <h2>Expense List</h2>
      {
        props.expenses.map((expense) => {
          return <ExpenseListItem key={expense.id} {...expense} />;
        })
      }
    </div>
  );
};

ExpenseList.propTypes = requiredProps;

const mapStateToProps = (state) => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
