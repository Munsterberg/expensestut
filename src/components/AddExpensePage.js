import React from 'react';
import { connect } from 'react-redux';
import { func, object } from 'prop-types';

import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    static propTypes = {
      dispatch: func,
      history: object,
      addExpense: func
    }

  _addExpense = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  }
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this._addExpense} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addExpense: (expense) => dispatch(addExpense(expense))
  };
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
