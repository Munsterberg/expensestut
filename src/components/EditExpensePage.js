import React from 'react';
import { object, func } from 'prop-types';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  static propTypes = {
    match: object,
    expense: object,
    dispatch: func,
    history: object,
    onEditExpense: func,
    onRemoveExpense: func
  }

  _onSubmit = (expense) => {
    this.props.onEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }

  _onRemoveExpense = () => {
    this.props.onRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <h1>Edit Expense: {this.props.expense.id}</h1>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this._onSubmit} />
        <button onClick={this._onRemoveExpense}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    onEditExpense: (expenseId, expense) => dispatch(editExpense(expenseId, expense)),
    onRemoveExpense: (expense) => dispatch(removeExpense(expense))
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);