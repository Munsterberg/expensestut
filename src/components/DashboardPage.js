import React from 'react';

import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

function DashboardPage() {
  return (
    <div>
      <ExpenseListFilters />
      <ExpenseList />
    </div>
  );
}

export default DashboardPage;
