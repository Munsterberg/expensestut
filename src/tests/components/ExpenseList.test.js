import React from 'react';
import { shallow } from 'enzyme';

import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

let wrapper;

beforeEach(() => {
  wrapper = shallow(<ExpenseList expenses={expenses} />);
});

test('should render ExpenseList with expenses', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should be type div', () => {
  expect(wrapper.type()).toBe('div');
});

test('should render 3 ExpenseListItem components', () => {
  expect(wrapper.find('ExpenseListItem')).toHaveLength(3);
});