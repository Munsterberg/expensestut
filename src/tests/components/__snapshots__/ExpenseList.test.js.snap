// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <h2>
    Expense List
  </h2>
  <ExpenseListItem
    amount={185}
    createdAt={0}
    description="Water bill"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={20000}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={4000}
    createdAt={345600000}
    description="Credit Card "
    id="3"
    key="3"
    note=""
  />
</div>
`;
